hal/network/ncp/wifi/platform_ncp_update.cpp

```diff
@@ -0,0 +1,167 @@
+        }
+        size = std::min(size, remaining);
+        buffer += size;
+        remaining -= size;
+        LED_Toggle(LED_RGB);
+        return size;
+    }
+
+    int availForRead() override {
+        return remaining;
+    }
+
+    int waitEvent(unsigned flags, unsigned timeout) override {
+        if (!flags) {
+            return 0;
+        }
+        if (!(flags & InputStream::READABLE)) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        if (!remaining) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        return InputStream::READABLE;
+    }
+
+private:
+    const uint8_t* buffer;
+    size_t remaining;
+};
+
+// FIXME: This function accesses the module info via XIP and may fail to parse it correctly under
+// some not entirely clear circumstances. Disabling compiler optimizations helps to work around
+// the problem
+__attribute__((optimize("O0"))) int platform_ncp_update_module(const hal_module_t* module) {
+    const auto ncpClient = particle::wifiNetworkManager()->ncpClient();
+    SPARK_ASSERT(ncpClient);
+    CHECK(ncpClient->on());
+    // we pass only the actual binary after the module info and up to the suffix
+    const uint8_t* start = (const uint8_t*)module->info;
+    static_assert(sizeof(module_info_t)==24, "expected module info size to be 24");
+    start += sizeof(module_info_t); // skip the module info
+    const uint8_t* end = start + (uint32_t(module->info->module_end_address) - uint32_t(module->info->module_start_address));
+    const unsigned length = end-start;
+    OtaUpdateSourceStream moduleStream(start, length);
+    uint16_t version = 0;
+    int r = ncpClient->getFirmwareModuleVersion(&version);
+    if (r == 0) {
+        LOG(INFO, "Updating ESP32 firmware from version %d to version %d", version, module->info->module_version);
+    }
+    r = ncpClient->updateFirmware(&moduleStream, length);
+    LED_On(LED_RGB);
+    CHECK(r);
+    r = ncpClient->getFirmwareModuleVersion(&version);
+    if (r == 0) {
+        LOG(INFO, "ESP32 firmware version updated to version %d", version);
+    }
+    return HAL_UPDATE_APPLIED;
+}
+
+int platform_ncp_fetch_module_info(hal_system_info_t* sys_info, bool create) {
+    for (int i=0; i<sys_info->module_count; i++) {
+        hal_module_t* module = sys_info->modules + i;
+        if (!memcmp(&module->bounds, &module_ncp_mono, sizeof(module_ncp_mono))) {
+            if (create) {
+                const auto ncpClient = particle::wifiNetworkManager()->ncpClient();
+                SPARK_ASSERT(ncpClient);
+                CHECK(ncpClient->on());
+                uint16_t version;
+                int error = ncpClient->getFirmwareModuleVersion(&version);
+                if (error) {
+                    version = 0;
+                }
+                // todo - we could augment the getFirmwareModuleVersion command to retrieve more details
+                auto info = new module_info_t();
```

Comment:
nothrow?

Reply:
There is also no difference between `new` and `new (std::nothrow)` on current platforms other than stylistic as exceptions are disabled and there is a check below. We can add it, but again https://github.com/particle-iot/device-os/pull/2250#discussion_r539603067

Reply:
Up to you for consistency

Reply:
Then I'll leave it as-is, since this code block was simply moved and not a new change. We don't have a strong policy on `new (std::nothrow)` either.